[Upgrade Assistant] Fix isSuccessful check in verifySnapshotUpgrade method

sabarasaba opened the pull request:
# Summary

When we attempt to update a ML snapshot model we have a method thast checks if the upgrade was successful. If we have only one deprecation the [first condition](https://github.com/elastic/kibana/blob/7.17/x-pack/plugins/upgrade_assistant/server/routes/ml_snapshots.ts#L73) from this method will succeed and it will return that the snapshot was migrated correctly. But if we have more than one snapshot to be migrated this condition doesn't run and instead it will [attempt to find](https://github.com/elastic/kibana/blob/7.17/x-pack/plugins/upgrade_assistant/server/routes/ml_snapshots.ts#L79) in the `mlSnapshotDeprecations` list if we have a match for the snapshot we're updating and nothing is found then the update was successful.

The problem was that there was some alterations made to the format of the deprecation titles and now it has 3 sets of brackets instead of two, so the UI was reporting that the upgrade failed while in reality it succeeded.

**Steps to test:**
1. Run elasticsearch and kibana
2. Create and run **TWO** really simple ML anomaly detection jobs. This can be run on Kibana sample data, for example the sample flight data, and can be as simple as possible, for example a single metric count job.  Just run the job on the historical data, don't ask for it to continue in real-time.
3. Go to dev tools and update the model snapshot definition docs to say they came from 6.3:

```
GET .ml-anomalies-shared/_search
{
  "query": {
    "exists": { "field" : "snapshot_id" }
  }
}

POST .ml-anomalies-shared/_update_by_query
{
  "query": {
    "exists": { "field" : "snapshot_id" }
  },
  "script": {
    "source": "ctx._source.min_version='6.3.0'",
    "lang": "painless"
  }
}

GET .ml-anomalies-shared/_search
{
  "query": {
    "exists": { "field" : "snapshot_id" }
  }
}
```

4. Now go to upgrade assistant and look at the deprecation warnings.  The hack in step 4 should convince it that the model snapshot created by the simple job created in step 3 needs upgrading.
5. Make sure that both ML deprecations can be addressed correctly.
6. To iterate, steps 4 and 5 can be repeated as many times as it takes.

elasticmachine commented:
Pinging @elastic/platform-deployment-management (Team:Deployment Management)

sabarasaba commented on x-pack/plugins/upgrade_assistant/server/routes/ml_snapshots.ts:
```diff
@@ -310,6 +314,9 @@ export function registerMlSnapshotRoutes({ router, lib: { handleEsError } }: Rou
               });
             }
 
+            log.error(
```
Since both `customError` messages are the same in this route, I've added an extra error log so that if this error ever comes up, we could tell from which of the two branches it came from and what response caused it. This shouldn't happen if everything worked successfully, so it shouldn't cause log spam for users and will make it easier for us to debug this kind of problems.

droberts195 reviewed (commented):
I think there's still room to make the messages more specific.  It might not make much difference to the end user, but making the following changes will help ML developers who have to deal with support cases for this functionality in the future.

Please could you change the message on line 263 from:

> The upgrade that was started for this model snapshot doesn't exist anymore.

to:

> The upgrade process for this model snapshot failed. Check the Elasticsearch logs for more details.

And also, please change the message on line 293 from:

> There was an error upgrading your snapshot. Check the Elasticsearch logs for more details.

to:

> The upgrade process for this model snapshot stopped yet the snapshot is not upgraded. Check the Elasticsearch logs for more details.

And change the message on line 325 from:

> There was an error upgrading your snapshot. Check the Elasticsearch logs for more details.

to:

> The upgrade process for this model snapshot completed yet the snapshot is not upgraded. Check the Elasticsearch logs for more details.

None of these messages should be seen by the end user if everything works as intended, but if they are then being precise about what happened will aid debugging.

sabarasaba commented:
Thanks for the suggestion @droberts195! I've addressed your comments with https://github.com/elastic/kibana/pull/122541/commits/f1b0d83b4b8329324aa81cb58e4b6092d544a54d

droberts195 reviewed (approved):
LGTM - thanks!

sebelga commented on x-pack/plugins/upgrade_assistant/server/routes/ml_snapshots.ts:
```diff
@@ -81,7 +81,7 @@ const verifySnapshotUpgrade = async (
         const matches = snapshotDeprecation.message.match(regex);
 
-        if (matches?.length === 2) {
+        if (matches?.length === 3) {
```
I know it is not part of this PR but any chance we could add a comment on top of the `regex` to explain exactly what we are looking for. Then it would be easier to understand why we needed to change this line.

sebelga commented on x-pack/plugins/upgrade_assistant/server/routes/ml_snapshots.ts:
```diff
@@ -256,7 +260,7 @@ export function registerMlSnapshotRoutes({ router, lib: { handleEsError } }: Rou
                     {
                       defaultMessage:
-                        "The upgrade that was started for this model snapshot doesn't exist anymore.",
+                        'The upgrade process for this model snapshot failed. Check the Elasticsearch logs for more details.',
```
I wonder if it wouldn't be good to add metadata with the error message and the `log()` call you added below.

Something like

```
`The upgrade process for this model [${modelId}] failed...`
```

And maybe the snapshot id, job id if it make sense to include them?

sabarasaba replied:
Thanks for the review! I don't think it adds of much since we will know which job/snapshot failed after kicking the upgrade since the UI and the api call will tell us all we need

cjcenizal replied:
If an admin is looking at the logs of a large deployment I think it's unlikely to have access to the UI or the original API call at the time the problem occurs. Would this info be useful in that case?

sabarasaba replied:
That's a good point @cjcenizal! Ill patch it up in a bit!

sebelga replied:
Yes that's what I thought. If there is an SDH case coming up we'll only have the logs to look at. The more there is there the better 😊

sebelga reviewed (approved):
LGTM! Tested locally and works as expected. Great job @sabarasaba !

I left a comment about adding metadata with the error messages returned to the user and the log message we output. I let you decide if it makes sense to add some more information to debug.

sabarasaba commented:
@elasticmachine merge upstream
